File: daily_problems/2025/01/0120/personal_submission/cf182e_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/convolution"
// #include "atcoder/dsu"
// #include "atcoder/fenwicktree"
// #include "atcoder/lazysegtree"
// #include "atcoder/math"
// #include "atcoder/maxflow"
// #include "atcoder/mincostflow"
// #include "atcoder/modint"
// #include "atcoder/scc"
// #include "atcoder/segtree"
// #include "atcoder/string"
// #include "atcoder/twosat"
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divide, I conquer!

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n, l;
    cin >> n >> l;

    vector<pair<int, int>> rects(n);
    for (auto &[x, y]: rects) cin >> x >> y;

    vector<vector<int>> dp(2 * n + 1, vector<int>(l + 1, 0));
    dp[2 * n][0] = 1;

    int mod = 1e9 + 7;

    auto add = [&] (int &x, int y) -> void {
        x += y;
        if (x >= mod) x -= mod;
    };

    for (int i = 0; i <= l; i ++) {
        for (int old = 0; old <= 2 * n; old ++) {
            for (int idx = 0; idx < n; idx ++) {
                if (old / 2 != idx) {
                    int v = (old == 2 * n ? 0 : (old & 1 ? rects[old / 2].second : rects[old / 2].first));

                    if ((v == 0 || rects[idx].first == v) && i + rects[idx].first <= l)
                        add(dp[idx * 2 + 1][i + rects[idx].first], dp[old][i]);
                    if (rects[idx].first != rects[idx].second && (v == 0 || rects[idx].second == v) && i + rects[idx].second <= l)
                        add(dp[idx * 2][i + rects[idx].second], dp[old][i]);
                }
            }
        }
    }

    int ans = 0;
    for (int i = 0; i <= 2 * n; i ++)
        add(ans, dp[i][l]);
    cout << ans;

    return 0;
}

File: daily_problems/2025/01/0120/personal_submission/cf182e_yawn_sean.py
# Submission link: https://codeforces.com/contest/182/submission/301928869
def main():
    n, l = MII()
    xs = []
    ys = []

    for _ in range(n):
        x, y = MII()
        xs.append(x)
        ys.append(y)

    dp = [[0] * (l + 1) for _ in range(n * 2 + 1)]
    dp[2 * n][0] = 1

    mod = 10 ** 9 + 7

    for i in range(l + 1):
        for old in range(2 * n + 1):
            for idx in range(n):
                if old // 2 != idx:
                    v = 0 if old == 2 * n else (ys[old // 2] if old % 2 else xs[old // 2])
                    if (v == 0 or xs[idx] == v) and i + xs[idx] <= l:
                        dp[idx * 2 + 1][i + xs[idx]] += dp[old][i]
                        dp[idx * 2 + 1][i + xs[idx]] %= mod
                    if xs[idx] != ys[idx] and (v == 0 or ys[idx] == v) and i + ys[idx] <= l:
                        dp[idx * 2][i + ys[idx]] += dp[old][i]
                        dp[idx * 2][i + ys[idx]] %= mod

    print(sum(dp[i][l] for i in range(2 * n + 1)) % mod)